### PR TITLE
[GlobalIsel] Canonicalize G_ICMP

### PR DESCRIPTION
**Pull request by tschuett**
As a side-effect, we start constant folding icmps.

Split out from https://github.com/llvm/llvm-project/pull/105991.

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+#include "llvm/IR/Instructions.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/ErrorHandling.h"
+#include <cstdlib>
+
+#define DEBUG_TYPE "gi-combiner"
+
+using namespace llvm;
+
+bool CombinerHelper::constantFoldICmp(const GICmp &ICmp,
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp:32]: Why not put this in the builder like other constant folds?

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp:32]: The builder only folds icmps that it builds. This PR adds one combine that matches G_ICMP. While running the combiner, we could find G_ICMPs where the parameters are constants due to combining. The builder would never build this G_ICMP again. In DAG terms, this is visitICmp.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+bool CombinerHelper::matchCanonicalizeICmp(const MachineInstr &MI,
+                                           BuildFnTy &MatchInfo) {
+  const GICmp *Cmp = cast<GICmp>(&MI);
+
+  Register Dst = Cmp->getReg(0);
+  Register LHS = Cmp->getLHSReg();
+  Register RHS = Cmp->getRHSReg();
+
+  CmpInst::Predicate Pred = Cmp->getCond();
+  assert(CmpInst::isIntPredicate(Pred) && "Not an integer compare!");
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp:72]: That would be a different PR. This PR has one pattern that matches G_ICMP. We would need a different pattern for G_FCMP. This PR introduces GIConstant, which is a wrapper around a vector of APInt. For G_FCMP, we would need to introduce GFConstant, which is a wrapper around a vector of APFloat.

This PR canonicalizes and constant folds icmps.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  if (auto CLHS = GIConstant::getConstant(LHS, MRI)) {
+    if (auto CRHS = GIConstant::getConstant(RHS, MRI))
+      return constantFoldICmp(*Cmp, *CLHS, *CRHS, MatchInfo);
+
+    // If we have a constant, make sure it is on the RHS.
+    std::swap(LHS, RHS);
+    Pred = CmpInst::getSwappedPredicate(Pred);
+
+    MatchInfo = [=](MachineIRBuilder &B) { B.buildICmp(Pred, Dst, LHS, RHS); };
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp:81]: I never do that!

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp:72]: This is a split out from https://github.com/llvm/llvm-project/pull/105991. It only handles icmps. We could add the same for fcmps, but the market share is probably lower without having data.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp (hunk 1 above)
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp:32]: We already have ConstantFoldICmp, but I needed this on top of GIConstants.
https://github.com/llvm/llvm-project/blob/76b54df87a860c20ac68e4ed1d766b545d412c1e/llvm/lib/CodeGen/GlobalISel/Utils.cpp#L1013

This used by the builder.

File: llvm/include/llvm/CodeGen/GlobalISel/Utils.h [hunk 4]
```diff
@@ -593,5 +593,27 @@ bool isGuaranteedNotToBeUndef(Register Reg, const MachineRegisterInfo &MRI,
 /// estimate of the type.
 Type *getTypeForLLT(LLT Ty, LLVMContext &C);
 
+enum class GIConstantKind { Scalar, FixedVector, ScalableVector };
+
+/// An integer-like constant.
```
Comment (aemerson) [inline llvm/include/llvm/CodeGen/GlobalISel/Utils.h:598]: This class could do with more documentation about its intended purpose, in what situation it's supposed to be used.

File: llvm/include/llvm/CodeGen/GlobalISel/Utils.h [hunk 5]
```diff
@@ -593,5 +593,27 @@ bool isGuaranteedNotToBeUndef(Register Reg, const MachineRegisterInfo &MRI,
 /// estimate of the type.
 Type *getTypeForLLT(LLT Ty, LLVMContext &C);
 
+enum class GIConstantKind { Scalar, FixedVector, ScalableVector };
```
Comment (aemerson) [inline llvm/include/llvm/CodeGen/GlobalISel/Utils.h:596]: Move this inside the class?

File: llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h [hunk 6]
```diff
@@ -950,6 +950,16 @@ class GExtOrTruncOp : public GCastOp {
   };
 };
 
+/// Represents a splat vector.
+class GSplatVector : public GenericMachineInstr {
+public:
+  Register getValueReg() const { return getOperand(1).getReg(); }
```
Comment (aemerson) [inline llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h:956]: To be more descriptive how about call this `getScalarReg()` instead?

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-visit-icmp.mir
Comment (aemerson) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-visit-icmp.mir:1]: This still needs renaming.

File: llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h (hunk 6 above)
Comment (tschuett) [reply llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h:956]:
```
// Generic splatvector.
def G_SPLAT_VECTOR: GenericInstruction {
  let OutOperandList = (outs type0:$dst);
  let InOperandList = (ins type1:$val);
  let hasSideEffects = false;
}
```
In the .td file it is called `val`.  Scalar is confusing for me on a scalable vector.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelperCompares.cpp:81]: Agreed! But I don't like this low-level manipulation of instructions. It is error prone. I prefer the build and erase pattern.

**Review by aemerson — APPROVED** [review]
LGTM, but I'm reserving the right to change the matchers to mutate the instruction in future.

**Pull request merged**